### PR TITLE
MODLD-395: inventory kafka sender for Instance Creation

### PR DESCRIPTION
**Pull request by PBobylev**

### CODE REVIEW

File: src/main/java/org/folio/linked/data/integration/ResourceModificationEventListener.java [hunk 1]
```diff
@@ -51,4 +61,20 @@ public void afterIndex(ResourceIndexedEvent resourceIndexedEvent) {
     resourceRepository.updateIndexDate(resourceIndexedEvent.workId());
   }
 
+  private void sendToSearch(Resource resource) {
+    if (isOfType(resource, WORK)) {
+      kafkaSearchSender.sendSingleResourceCreated(resource);
+    } else {
+      extractWork(resource)
+        .map(Resource::getId)
```
Comment (pkjacob) [inline src/main/java/org/folio/linked/data/integration/ResourceModificationEventListener.java:69]: Hi @PBobylev,

Do we need  these 2 lines here?

```
        .map(Resource::getId)
        .map(resourceRepository::getReferenceById)
```

We already have `Optional<Work>` in line 68.  (Which is actually retrieved in line 40). No point in fetching the same from DB again right?

Comment (PBobylev) [reply src/main/java/org/folio/linked/data/integration/ResourceModificationEventListener.java:69]: right, changed

File: src/main/java/org/folio/linked/data/integration/kafka/sender/search/KafkaSearchSenderImpl.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    return kafkaSearchMessageMapper.toIndex(resource, CREATE)
+      .map(bibframeIndex -> {
+        sendCreate(bibframeIndex, false);
+        return true;
+      })
+      .isPresent();
+  }
+
+  @SneakyThrows
+  private void sendCreate(BibframeIndex bibframeIndex, boolean publishIndexEvent) {
```
Comment (aleksei-pronichev) [inline src/main/java/org/folio/linked/data/integration/kafka/sender/search/KafkaSearchSenderImpl.java:53]: minor
using `if boolean parameter` isn't good practice
suggest to make separate methods
```java
private void sendCreateWithPublishingIndex(BibframeIndex bibframeIndex) {
    sendCreate(bibframeIndex)
   eventPublisher.publishEvent(new ResourceIndexedEvent(parseLong(bibframeIndex.getId())));
}

private void sendCreate(BibframeIndex bibframeIndex) {
    var message = new SearchIndexEvent()
      .id(UUID.randomUUID().toString())
      .type(CREATE)
      .resourceName(SEARCH_RESOURCE_NAME)
      ._new(bibframeIndex);
    searchIndexEventMessageProducer.sendMessages(List.of(message));
}

```

**Review by pkjacob — APPROVED** [review]

File: src/main/java/org/folio/linked/data/integration/kafka/sender/search/KafkaSearchSenderImpl.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+@Service
+@Profile(FOLIO_PROFILE)
+@RequiredArgsConstructor
+public class KafkaSearchSenderImpl implements KafkaSearchSender {
+
+  private final ApplicationEventPublisher eventPublisher;
+  private final KafkaSearchMessageMapper kafkaSearchMessageMapper;
+  private final FolioMessageProducer<SearchIndexEvent> searchIndexEventMessageProducer;
+
+  @SneakyThrows
```
[the remark this answers is not in the document]
Comment (PBobylev) [reply src/main/java/org/folio/linked/data/integration/kafka/sender/search/KafkaSearchSenderImpl.java:36]: it's not required anymore, removed, thx!

File: src/main/java/org/folio/linked/data/integration/kafka/sender/search/KafkaSearchSenderImpl.java (hunk 2 above)
Comment (PBobylev) [reply src/main/java/org/folio/linked/data/integration/kafka/sender/search/KafkaSearchSenderImpl.java:53]: agree, and we could make it even more optimal just separating methods. Changed, thx!

File: src/main/java/org/folio/linked/data/util/BibframeUtils.java [hunk 4]
```diff
@@ -78,4 +82,28 @@ public static boolean isSameResource(BibframeIndex index, Resource resource) {
     }
   }
 
+  public static Optional<Resource> extractWork(Resource resource) {
+    return isOfType(resource, WORK) ? Optional.of(resource)
```
[the remark this answers is not in the document]
Comment (PBobylev) [reply src/main/java/org/folio/linked/data/util/BibframeUtils.java:86]: sure, done, thx!

**Review by askhat-abishev — APPROVED** [review]

File: src/main/java/org/folio/linked/data/integration/kafka/sender/inventory/KafkaInventorySenderImpl.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import lombok.SneakyThrows;
+import lombok.extern.log4j.Log4j2;
+import org.folio.linked.data.mapper.kafka.KafkaInventoryMessageMapper;
+import org.folio.linked.data.model.entity.Resource;
+import org.folio.search.domain.dto.InstanceIngressEvent;
+import org.folio.spring.tools.kafka.FolioMessageProducer;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+
+@Log4j2
```
Comment (askhat-abishev) [inline src/main/java/org/folio/linked/data/integration/kafka/sender/inventory/KafkaInventorySenderImpl.java:17]: not used

File: src/main/java/org/folio/linked/data/mapper/kafka/impl/KafkaInventoryMessageMapperImpl.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.linked.data.mapper.ResourceModelMapper;
+import org.folio.linked.data.mapper.kafka.KafkaInventoryMessageMapper;
+import org.folio.linked.data.model.entity.Resource;
+import org.folio.marc4ld.service.ld2marc.Bibframe2MarcMapper;
+import org.folio.search.domain.dto.InstanceIngressPayload;
+import org.springframework.stereotype.Component;
+
+@Log4j2
```
Comment (askhat-abishev) [inline src/main/java/org/folio/linked/data/mapper/kafka/impl/KafkaInventoryMessageMapperImpl.java:16]: not used

File: src/main/java/org/folio/linked/data/integration/kafka/sender/inventory/KafkaInventorySenderImpl.java (hunk 5 above)
Comment (PBobylev) [reply src/main/java/org/folio/linked/data/integration/kafka/sender/inventory/KafkaInventorySenderImpl.java:17]: removed, thx!

File: src/main/java/org/folio/linked/data/mapper/kafka/impl/KafkaInventoryMessageMapperImpl.java (hunk 6 above)
Comment (PBobylev) [reply src/main/java/org/folio/linked/data/mapper/kafka/impl/KafkaInventoryMessageMapperImpl.java:16]: removed, thx!

**Pull request merged**